x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java

```diff
@@ -10,15 +10,26 @@
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.core.AbstractRefCounted;
 import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
 import org.elasticsearch.transport.TransportResponse;
 
 import java.io.IOException;
 import java.util.Objects;
 
-public final class ExchangeResponse extends TransportResponse {
+public final class ExchangeResponse extends TransportResponse implements Releasable {
+    private final RefCounted counted = AbstractRefCounted.of(this::close);
     private final Page page;
     private final boolean finished;
+    /**
+     * We always use the remote exchange framwork even for local exchanges, but
+     * local exchanges shouldn't close the Page. Remote exchanges totally should
+     * as soon as they've serialized the block. This is a clever hack Nhat mentioned
+     * that can do that. But I don't like it. But it works and might unstick us.
```

Comment:
I think that this is fine.

Reply:
:+1:

@dnhatn, if you have the time to remove this hack then please be my guest, but we might live with it for a few months.

Reply:
I've opened: https://github.com/elastic/elasticsearch/pull/99891